pkg/virt-api/webhooks/arm64.go

```diff
@@ -0,0 +1,156 @@
+ /* Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2019 Red Hat, Inc.
+ *
+ */
+
+/*
+ * arm64 utilities are in the webhooks package because they are used both
+ * by validation and mutation webhooks.
+ */
+package webhooks
+
+import (
+        "fmt"
+	rt "runtime"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sfield "k8s.io/apimachinery/pkg/util/validation/field"
+
+        v1 "kubevirt.io/client-go/api/v1"
+)
+
+var _false bool = false
+
+const (
+	defaultCPUModel = "host-passthrough"
+)
+
+// verifyInvalidSetting verify if VMI spec contain unavailable setting for arm64, check following items:
+// 1. if setting bios boot
+// 2. if use uefi secure boot
+// 3. if set auto-attach graphics device
+// 4. if use host-model for cpu model
+func verifyInvalidSetting(field *k8sfield.Path, spec *v1.VirtualMachineInstanceSpec) (metav1.StatusCause, bool) {
+	if spec.Domain.Firmware != nil && spec.Domain.Firmware.Bootloader !=nil {
+		if spec.Domain.Firmware.Bootloader.BIOS != nil {
+			return metav1.StatusCause{
+				Type:		metav1.CauseTypeFieldValueNotSupported,
+				Message:	"Arm64 does not support bios boot, please change to uefi boot",
+				Field:		field.Child("domain", "firmware", "bootloader", "bios").String(),
+			} , false
+		}
+		if spec.Domain.Firmware.Bootloader.EFI != nil {
+			// When EFI is enable, secureboot is enabled by default, so here check two condition
+			// 1 is EFI is enabled without Secureboot setting
+			// 2 is both EFI and Secureboot enabled
+			if spec.Domain.Firmware.Bootloader.EFI.SecureBoot == nil || (spec.Domain.Firmware.Bootloader.EFI.SecureBoot != nil && *spec.Domain.Firmware.Bootloader.EFI.SecureBoot) {
```

Comment:
> Can this happen? and is that allowed on ARM if this is not set?

Yes, if `spec.Domain.Firmware.Bootloader.EFI.SecureBoot == nil` which means it would use EFI default method to boot VM and the default method is SecureBoot, which is not allowed on ARM64;

if `spec.Domain.Firmware.Bootloader.EFI.SecureBoot != nil && *spec.Domain.Firmware.Bootloader.EFI.SecureBoot` which means user do want use UEFI secure boot, which is also not allowed on ARM64.

Here are two condition which is allowed on ARM,
`spec.Domain.Firmware.Bootloader == nil` is not set or `spec.Domain.Firmware.Bootloader.EFI == true && !*spec.Domain.Firmware.Bootloader.EFI.SecureBoot`

Reply:
Would it make sense to just set it by default if it is nil? We normally do not require that people set fields in the bootloader section. It makes perfect sense to reject invalid configurations, but picking the right default would work too for me.

Reply:
> Would it make sense to just set it by default if it is nil? 

I am not sure which variable the "it" refer to, "spec.Domain.Firmware.Bootloader" or "spec.Domain.Firmware.Bootloader.EFI" or "spec.Domain.Firmware.Bootloader.EFI.SecureBoot"?

> We normally do not require that people set fields in the bootloader section. It makes perfect sense to reject invalid configurations, but picking the right default would work too for me.

If `spec.Domain.Firmware.Bootloader == nil` and architecture is arm64, we will set correct bootloader for user in following function `setDefaultBootloader`

Reply:
Perfect. :+1: